sound/soc/sof/ipc4-topology.c

```diff
@@ -427,6 +427,70 @@ static void sof_ipc4_widget_free_comp_pcm(struct snd_sof_widget *swidget)
 	swidget->private = NULL;
 }
 
+static int sof_ipc4_widget_setup_comp_buffer(struct snd_sof_widget *swidget)
+{
+	struct sof_ipc4_available_audio_format *available_fmt;
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct sof_ipc4_copier *ipc4_copier;
+	int ret;
+
+	ipc4_copier = kzalloc(sizeof(*ipc4_copier), GFP_KERNEL);
+	if (!ipc4_copier)
+		return -ENOMEM;
+
+	swidget->private = ipc4_copier;
+	available_fmt = &ipc4_copier->available_fmt;
+
+	dev_dbg(scomp->dev, "Updating IPC structure for %s\n", swidget->widget->name);
+
+	ret = sof_ipc4_get_audio_fmt(scomp, swidget, available_fmt, true);
+	if (ret)
+		goto free_copier;
+
+	available_fmt->dma_buffer_size = kcalloc(available_fmt->audio_fmt_num, sizeof(u32),
+						 GFP_KERNEL);
+	if (!available_fmt->dma_buffer_size) {
+		ret = -ENOMEM;
+		goto free_available_fmt;
+	}
+
+	/* set invalid gtw_cfg for module type copier */
+	ipc4_copier->data.gtw_cfg.node_id = -1;
```

Comment:
-1? Is this what is ecptected by the FW for copier module?

Reply:
> -1? Is this what is ecptected by the FW for copier module?

This was suggested by @RanderWang. I tested with SOF and it works. But it doesn't work with IPC4 tester. I don't know what IPC4 tester expected.

Reply:
define a macro like FW #define IPC4_INVALID_NODE_ID	0xffffffff

Reply:
> define a macro like FW #define IPC4_INVALID_NODE_ID 0xffffffff

Done